amplify/backend/function/getMembersForColony/src/index.js

```diff
@@ -121,177 +121,142 @@ exports.handler = async (event) => {
+
   // Incase there are members but no reputation on the colony,
   // we can stop here and return the list of watchers
-  if (!addressesWithReputation?.length) {
+  if (!contributorAddresses.size) {
+    if (domainId > Id.RootDomain) {
+      return {
+        contributors: [],
+        watchers: [], // You can only "watch" the root domain.
+      };
+    }
+
     return {
       contributors: [],
-      watchers:
-        domainId > Id.RootDomain
-          ? [] // There will be no Watchers outside of the root domain
-          : data?.getColonyByAddress?.items[0]?.watchers.items.map(
-              ({ user }) => ({
-                address: user.id,
-                user,
-              }),
-            ),
+      watchers: allColonyWatchers.map(({ user }) => ({
+        address: user.id,
+        user,
+      })),
     };
   }
 
-  // Identify watchers & contributors
-  data?.getColonyByAddress?.items[0]?.watchers.items.forEach((item) => {
-    if (
-      addressesWithReputation?.some(
-        (address) => address.toLowerCase() === item.user.id.toLowerCase(),
-      )
-    ) {
-      contributors.push({
-        address: item.user.id,
-        user: item.user,
-      });
-    } else {
-      watchers.push({
-        address: item.user.id,
-        user: item.user,
-      });
-    }
-  });
-
-  // There will be no Watchers outside of the root domain
-  if (domainId > Id.RootDomain) {
-    watchers.length = 0;
-  }
-
-  // now catch addresses that have reputation but are not in the watchers list
-  // i.e. address that was awarded reputation but has not joined the colony
-  if (addressesWithReputation?.length !== contributors?.length) {
-    const missingAddresses = addressesWithReputation?.filter((address) =>
-      contributors?.every(
-        (item) => item.user.id.toLowerCase() !== address.toLowerCase(),
-      ),
-    );
-
-    missingAddresses?.forEach((address) => {
-      contributors?.push({
-        address,
-      });
-    });
-  }
-
   // Get total reputation for colony
-  let totalColonyReputation;
+  let reputationInDomain;
   try {
-    totalColonyReputation = await colonyClient.getReputationWithoutProofs(
+    const { reputationAmount } = await colonyClient.getReputationWithoutProofs(
       skillId,
       constants.AddressZero,
       rootHash,
     );
+    reputationInDomain = reputationAmount;
   } catch (error) {
-    // Not throwing anything, as its possible that the domain does not have
-    // any reputation, and we don't want to break the whole query
+    // If domain has no reputation, either it's the root domain, in which case we've already returned early above
+    // Or it's a subdomain, in which case we can return two empty arrays since you can't watch a subdomain
+    return {
+      contributors: [],
+      watchers: [],
+    };
   }
 
-  // get reputation for each address
-  const contributorsWithReputation = await Promise.all(
-    contributors.map(async (contributor) => {
```

Comment:
This logic was hard to follow, but basically it handles calculating the user's reputation in the domain. I have simplified.